### ISSUE TITLE
UnauthorizedAccessException when building .net core 6.0 project on MacOS

### ISSUE THREAD

**Issue by Joseph-Anthony-King** [dotnet/aspnetcore#41307]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have dotnet core API that I'm developing on MacOS.  It's been working great but recent updates to the API are causing the build to fail with an UnauthorizedAccessException for the API dll in the debug file.  The error is as follows:

```
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(5,5): Error: UnauthorizedAccessException: Access to the path '/Users/jking/Source/repos/SudokuCollective/SudokuCollective.Api/obj/Debug/net6.0/refint/SudokuCollective.Api.dll' is denied.
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Interop.CheckIo(Error error, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at Microsoft.CodeAnalysis.BuildTasks.CopyRefAssembly.Copy()
IOException: Operation not permitted

 (SudokuCollective.Api)
```

When I rollback to an earlier state I can build the project.  I'm able to build and run the project on windows, this seems to be  an issue when trying to build this project on MacOS.

### Expected Behavior

The project should build as it does on windows.

### Steps To Reproduce

1. Download the project located at: https://github.com/Joseph-Anthony-King/SudokuCollective
2. Use the dummySettings.json file to set up the appsettings.json file
3. Ensure that Postgres is running on your system
4. Run dotnet build SudokuCollective.sln

### Exceptions (if any)

```
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(5,5): Error: UnauthorizedAccessException: Access to the path '/Users/jking/Source/repos/SudokuCollective/SudokuCollective.Api/obj/Debug/net6.0/refint/SudokuCollective.Api.dll' is denied.
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Interop.CheckIo(Error error, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at Microsoft.CodeAnalysis.BuildTasks.CopyRefAssembly.Copy()
IOException: Operation not permitted

 (SudokuCollective.Api)
```

### .NET Version

6.0.202

**Comment by Rick-Anderson** [comment]
@Joseph-Anthony-King  you need the simplest possible repo, not your real project. 
Can you post to a public GitHub repo the [simplest possible sample](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces this problem? Please read the link on [simplest possible sample](https://stackoverflow.com/help/minimal-reproducible-example)

**Comment by Joseph-Anthony-King** [comment]
Hi @Rick-Anderson , I'll try to pair this down to see if I can recreate it.  Before I added the AppsController and ControllerUtilities it was working fine..  This is actually a work machine, my own personal machine is a dual boot windows/linuxbox.  Even though I just upgraded to 6.0.202 I'll take it out and reinstall as well.  It works on my windows machine.

**Comment by Rick-Anderson** [comment]
See if you can reproduce it under https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-sandbox/windows-sandbox-overview

**Comment by Rick-Anderson** [comment]
@Joseph-Anthony-King just make another clone on your Mac and see if you can reproduce it. It seems to me this has nothing to do with ASP.NET Core. There's a slim chance it has to do with the C# compiler/linker on the Mac.

**Comment by Joseph-Anthony-King** [comment]
Hi @Rick-Anderson , I think you're right... it isn't an ASP.NET Core issue.  I created a second copy, cloned from GitHub, and ran into the same issue.  I deleted the newest controller, AppsController.cs, and was able to build the project.  Its the same error I posted before.  For some reason adding new controllers is causing the following error... but I think you're right and this may not be a .Net Core issue:

```
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error MSB3883: Unexpected exception:  [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error : UnauthorizedAccessException: Access to the path '/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/obj/Debug/net6.0/refint/SudokuCollective.Api.dll' is denied. [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter) [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at Interop.CheckIo(Error error, String path, Boolean isDirectory, Func`2 errorRewriter) [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode) [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize) [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite) [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite) [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :    at Microsoft.CodeAnalysis.BuildTasks.CopyRefAssembly.Copy() [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error : IOException: Operation not permitted [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :  [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
/usr/local/share/dotnet/sdk/6.0.202/Microsoft.Common.CurrentVersion.targets(4650,5): error :  [/Users/jking/Source/repos/clone/SudokuCollective/SudokuCollective.Api/SudokuCollective.Api.csproj]
    0 Warning(s)
    2 Error(s)
```

**Comment by Joseph-Anthony-King** [comment]
Hello everyone, anytime I add any file to the SudokuCollective.Api project, not just controllers, get the UnauthorizedAccessException.  The project works on windows.  This has to be a C# compiler issue on Mac... anytime I add a file I'm getting this error.  Where do I direct the issue to... I don't think think this is an ASP.NET Core issue.

**Comment by Rick-Anderson** [comment]
@Joseph-Anthony-King 

1. Reproduce the problem on a clean machine. My guess is your machine has a problem.
2. Create the [simplest possible sample](https://stackoverflow.com/help/minimal-reproducible-example) that reproduces this problem. If the previous step fails, this step usually finds the problem.
3. Once you've done the previous steps, create an issue at https://github.com/dotnet/roslyn/issues

Do you mind if I close this issue now?

**Comment by Joseph-Anthony-King** [comment]
No, that's appropriate @Rick-Anderson , thank you for your assistance

**Comment by Joseph-Anthony-King** [comment]
In case anyone comes across a similar issue on MacOS I found a solution, it's not elegant but it works.  The SudokuCollective.Api project was somehow corrupted.  Anytime I added a file to the project it would through an unauthorized access exception.  Basically, whenever the project is compiled the dll is uploaded to [project]/obj/debug/net6.0.  However, due to whatever corruption occurred in this project added files were causing an unauthorized access exception whenever the compiler tried to copy the new dll into [project]/obj/debug/net6.0.

To solve the issue I copied the SudokuCollective.Api file to the desktop and recreated it copying the contents from the desktop to the new project file by file.  Wasn't elegant, but it worked.
